### PR TITLE
feat: implement shiftRight function push down

### LINKED ISSUE #5100: Implement `RightShift` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `RightShift` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by Willendless**
### What problem does this PR solve?

Issue Number: close #5100 
related pr: [pingcap/tidb#35872](https://github.com/pingcap/tidb/pull/35872)

Problem Summary:

Implement rightShift function push down.

### What is changed and how it works?

1. Register `bitShiftRight` func name in `scalar_func_map`.
2. Wrap builder with `buildBitwiseFunction` by updating `DAGExpressionAnalyzerHelper::function_builder_map`.
3. Update CH version's `BitShiftRightImpl` to handle a C++ undefined behavior of shift operation.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support to pushdown shiftRight to TiFlash
```

### CODE REVIEW

**Comment by xzhangxian1008** [comment]
Looks good

File: dbms/src/Functions/tests/gtest_bitshiftright.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            std::numeric_limits<std::uint64_t>::max()
+    );
+    size_t count = 100;
+    std::vector<uint64_t> v1(count), v2(count), res(count);
+    for (size_t i=0; i<count; ++i) {
+        v1[i] = dis1(gen);
+        v2[i] = dis2(gen);
+        res[i] = 0;
+    }
+    */
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_bitshiftright.cpp:269]: Does this code generate the following case?

**Comment by SeaRise** [comment]
Add release note like 'Support to pushdown bitShiftRight to TiFlash'.
And add the pr link of tidb?

File: dbms/src/Functions/tests/gtest_bitshiftright.cpp (hunk 1 above)
Comment (Willendless) [reply dbms/src/Functions/tests/gtest_bitshiftright.cpp:269]: Yes it does 👍

File: dbms/src/Functions/tests/gtest_bitshiftright.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt8>>({1}), createColumn<UInt16>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt16>>({1}), createColumn<UInt32>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt32>>({1}), createColumn<UInt64>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt8>>({1}), createColumn<UInt64>({0}), createColumn<Nullable<UInt64>>({1}));
+
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<Int16>>({1}), createColumn<UInt32>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<Int64>>({1}), createColumn<UInt8>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt32>>({1}), createColumn<Int16>({0}), createColumn<Nullable<UInt64>>({1}));
+    ASSERT_BITSHIFTRIGHT(createColumn<Nullable<UInt8>>({1}), createColumn<Int64>({0}), createColumn<Nullable<UInt64>>({1}));
+}
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_bitshiftright.cpp:117]: add test only input null like `createColumn<Nullable<Int8>>({})`?

**Review by SeaRise — APPROVED** [review]
Others LGTM

File: dbms/src/Functions/tests/gtest_bitshiftright.cpp (hunk 2 above)
Comment (Willendless) [reply dbms/src/Functions/tests/gtest_bitshiftright.cpp:117]: Can we assume this situation has been covered by case 5 of `TestFunctionBitShiftRight.TypeCastWithConst`, namely when input type is ColumnConst<Nullable> and value is null?

in L154 of [gtest_bitshiftright.cpp](https://github.com/pingcap/tiflash/pull/5156/files/6d1930da04040eca16a326801b718ed7f5b85293#diff-265fe8c8e979197d9a15f41b245cbbed28ba0da5bb790ecb13fb408652c8a5a0R154)

```cpp
    ASSERT_BITSHIFTRIGHT(createConstColumn<Nullable<Int8>>(4, std::nullopt), createColumn<UInt64>({0, 1, 0, 1}), createConstColumn<Nullable<UInt64>>(4, std::nullopt));
    ASSERT_BITSHIFTRIGHT(createConstColumn<Nullable<Int8>>(4, std::nullopt), createColumn<Nullable<UInt64>>({0, 1, std::nullopt, std::nullopt}), createConstColumn<Nullable<UInt64>>(4, std::nullopt));
    ASSERT_BITSHIFTRIGHT(createConstColumn<Nullable<Int8>>(4, std::nullopt), createConstColumn<UInt64>(4, 0), createConstColumn<Nullable<UInt64>>(4, std::nullopt));
    ASSERT_BITSHIFTRIGHT(createConstColumn<Nullable<Int8>>(4, std::nullopt), createConstColumn<UInt64>(4, 0), createConstColumn<Nullable<UInt64>>(4, std::nullopt));
    ASSERT_BITSHIFTRIGHT(createConstColumn<Nullable<Int8>>(4, std::nullopt), createConstColumn<Nullable<UInt64>>(4, std::nullopt), createConstColumn<Nullable<UInt64>>(4, std::nullopt));
```

**Review by zanmato1984 — APPROVED** [review]
LGTM

File: dbms/src/Functions/tests/gtest_bitshiftright.cpp (hunk 2 above)
Comment (SeaRise) [reply dbms/src/Functions/tests/gtest_bitshiftright.cpp:111]: ok

**Review by SeaRise — APPROVED** [review]
